### ISSUE TITLE
Permission Denied for RHEL 7.1 client 

### ISSUE THREAD

**Issue by GeorgeOikonomou** [jeroennijhof/pam_tacplus#36]
Hello,

After installing the module in a RHEL 7.1 machine and applying the following configuration:
1. Add in the  /etc/pam.d/sshd the following line:
```
auth       include      tac_plus
```
2. Creating file /etc/pam.d/tac_plus:

```
# %PAM-1.0
   
auth       sufficient   /usr/local/lib/security/pam_tacplus.so  debug server=mytacacs-server-ip  secret=my-secret
account    sufficient   /usr/local/lib/security/pam_tacplus.so  debug server=mytacacs-server-ip secret=my-secret  service=system protocol=ssl
session    sufficient   /usr/local/lib/security/pam_tacplus.so  debug server=mytacacs-server-ip secret=my-secret  service=system protocol=ssl
```

I get the following error message, when trying ssh user@mytacacs-server-ip :

```
Oct  7 10:36:44 localhost sshd[27315]: tac_connect_single: connection to mytacacs-server-ip:49 failed: Permission denied
Oct  7 10:36:44 localhost PAM-tacplus[27315]: connection failed srv 0: Permission denied
Oct  7 10:36:44 localhost PAM-tacplus[27315]: no more servers to connect
```

I stopped the clients (RHEL 7.1) firewall but the problem persists.
Moreover I have a working TACACS+ client (RHEL 6.6) with the same configuration and everything is fine. So probably the server is well configured.

Is there any compatibility problem with Redhat 7.1 ? 
How can I further troubleshoot that issue ?

Thanks

**Comment by ulf-f** [comment]
Hi,
I see the exact same issue here with "CentOS Linux release 7.2.1511 (Core)".
Did you already find any workaround?
Thanks
ulf

**Comment by jeroennijhof** [comment]
Could it be that selinux is enforcing anything?

**Comment by ulf-f** [comment]
Hi,
I fixed it with:
`semanage permissive -a sshd_t`

Thanks
ulf

**Comment by jeroennijhof** [comment]
@GeorgeOikonomou can you confirm it as well?

**Comment by cbpsnetworking** [comment]
I tried this command `semanage permissive -a sshd_t` and now get a new error on CentOS7 - clean install.

~~~
Dec 28 10:11:20 rhel7-centos sshd[15953]: tac_authen_read: short reply header, read 0 of 12: Operation now in progress
Dec 28 10:11:21 rhel7-centos sshd[15953]: tac_authen_read: short reply header, read -1 of 12: Connection reset by peer
Dec 28 10:11:21 rhel7-centos PAM-tacplus[15953]: no more servers to connect
~~~

Before running `semanage permissive -a sshd_t` I was receiving the same error:

~~~
Dec 27 12:12:15 rhel7-centos sshd[1475]: tac_connect_single: connection to 172.16.1.115:49 failed: Permission denied
Dec 27 12:12:15 rhel7-centos PAM-tacplus[1475]: connection failed srv 0: Permission denied
Dec 27 12:12:15 rhel7-centos sshd[1475]: tac_connect_single: connection to 172.17.1.115:49 failed: Permission denied
Dec 27 12:12:15 rhel7-centos PAM-tacplus[1475]: connection failed srv 1: Permission denied
Dec 27 12:12:15 rhel7-centos PAM-tacplus[1475]: no more servers to connect
Dec 27 12:13:02 rhel7-centos sshd[1475]: tac_connect_single: connection to 172.16.1.115:49 failed: Permission denied
Dec 27 12:13:02 rhel7-centos PAM-tacplus[1475]: connection failed srv 0: Permission denied
Dec 27 12:13:02 rhel7-centos sshd[1475]: tac_connect_single: connection to 172.17.1.115:49 failed: Permission denied
Dec 27 12:13:02 rhel7-centos PAM-tacplus[1475]: connection failed srv 1: Permission denied
~~~

**Comment by pprindeville** [comment]
Okay, you're being a little scattershot.

Maybe it's best if we focus on one platform at a time.  Can we look at the Debian build issues first?

The CentOS thing looks like an `sexlinux` issue.  The easiest way to verify this is:

```
# setenforce Permissive
```

and repeat your tests.  If it works, there's a high probability that it's an `selinux` issue.

Then you can collect AVC info as:

```
# ausearch -m avc -ts today | audit2why
```

and attach that.

**Comment by cbpsnetworking** [comment]
You are correct - I am just very determined at getting a walk-through for clean-install of Debian/CentOS for `pam_tacplus`.

I actually now got CentOS7 working after we were working on Debian. This is exact steps for CentOS from clean installation - I realize most of this should be realized by any user and left out but here it is anyway:

~~~
# Install sudo from root.
$ yum install sudo
# Install autoconf & automake.
$ sudo yum install autoconf
$ sudo yum install automake
# add test user with no password
# Install semanage:
$ sudo yum install policycoreutils-python
# run command: to get rid of the permission issue in /var/log/secure
$ sudo  semanage permissive -a sshd_t
# install openssl-devel 
$ sudo yum install openssl-devel
$ git clone https://github.com/jeroennijhof/pam_tacplus/
$ cd pam_tacplus
$ ./auto.sh
$ ./configure && make && sudo make install
~~~

*add `auth include tacacs` at top of `/etc/pam.d/sshd`:*
*add file `/etc/pam.d/tacacs` with:*
~~~
#%PAM-1.0
    auth       sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret
    account    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret service=shell protocol=ssh
    session    sufficient   /usr/local/lib/security/pam_tacplus.so debug server=172.16.1.115 secret=reallysecret service=shell protocol=ssh
~~~

*check libraries in CentOS:*

~~~
[eraddatz@rhel7-centos ~]$ ldd /usr/local/lib/security/pam_tacplus.so
	linux-vdso.so.1 =>  (0x00007ffde29f8000)
	libtac.so.2 => /usr/local/lib/libtac.so.2 (0x00007ff546824000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007ff546619000)
	libcrypto.so.10 => /lib64/libcrypto.so.10 (0x00007ff5461b8000)
	libpam.so.0 => /lib64/libpam.so.0 (0x00007ff545fa9000)
	libc.so.6 => /lib64/libc.so.6 (0x00007ff545be5000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007ff5459e1000)
	libz.so.1 => /lib64/libz.so.1 (0x00007ff5457cb000)
	libaudit.so.1 => /lib64/libaudit.so.1 (0x00007ff5455a2000)
	/lib64/ld-linux-x86-64.so.2 (0x000055b4b688c000)
	libcap-ng.so.0 => /lib64/libcap-ng.so.0 (0x00007ff54539c000)
~~~

*run an ssh with username that has no password:*

~~~
$ ssh netadmin@2.2.2.2
netadmin@2.2.2.2's password: 
Last login: Thu Dec 28 13:24:42 2017 from 1.1.1.1
[netadmin@rhel7-centos ~]$ 
~~~

**Comment by pprindeville** [comment]
> `$ sudo  semanage permissive -a sshd_t`

No.  Please go back and read my directions on how to use `ausearch` and `audit2allow`.  We can use this to generate a policy module to install as part of the package.

Letting a package which is under active development run unconfined is asking for trouble.  Do **not** do this under any circumstances.

One typically puts all or part of a system into `permissive` mode only long enough to figure out what its requirements are and to express those in an installed policy module.

**Comment by cbpsnetworking** [comment]
@pprindeville - thank you. I removed it first with `semanage permissive -a sshd_t` and had to do a reboot. I was then back to where I received the error `connection failed srv 1: Permission denied`.

I attempted the `setenforce Permissive` and this let it work temporarily. I rebooted again to get it back to non-working state and ran the ausearch - attached.

[ausearch-pam-tacplus.txt](https://github.com/jeroennijhof/pam_tacplus/files/1593398/ausearch-pam-tacplus.txt)

**Comment by pprindeville** [comment]
> I see in the output it says to run setsebool -P nis_enabled 1 so I did run it and it does make it work. 

It might be enough to mention it in a README.

**Comment by pprindeville** [comment]
For attribution purposes, what's your name?

**Comment by cbpsnetworking** [comment]
@pprindeville - I am Alex Tomko. Not sure how much attribution I can claim here but thank you.

**Comment by pprindeville** [comment]
Addressed in PR #104. Closing this issue as resolved.
